paddle/framework/lod_tensor.cc

```diff
@@ -170,20 +172,147 @@ LoDAndOffset GetSubLoDAndAbsoluteOffset(const LoD& lod, size_t start_idx,
   return LoDAndOffset{sub_lod, {start_idx, end_idx}};
 }
 
-void AppendLoD(LoD* lod, const LoD& lod_length) {
+void AppendLoD(LoD *lod, const LoD &lod_length) {
   PADDLE_ENFORCE(
       lod->empty() || lod->size() == lod_length.size(),
       "The lod_length should has the same size with the appended lod.");
   if (lod->empty()) {
     *lod = LoD(lod_length.size(), std::vector<size_t>({0}));
   }
   for (size_t i = 0; i < lod->size(); ++i) {
-    auto& level = (*lod)[i];
+    auto &level = (*lod)[i];
     for (size_t len : lod_length[i]) {
       level.push_back(level.back() + len);
     }
   }
 }
 
+void SerializeToStream(std::ostream &os, const LoDTensor &tensor,
```

Comment:
@helinwang I think the serialize function can be put into the `tensor_util.h`. Which is a part of this PR. https://github.com/PaddlePaddle/Paddle/pull/5455
Did you change the viewpoint after reading the boost serialized reference manul?

Reply:
Thanks for asking! It's ok if all the tensor types we that support currently and in the future will be only the `LoDTensor` (so we don't need polymorphism). Otherwise I think having serialization as a interface that every tensor implements would be better.

Reply:
Putting non-member functions in separated source files seems good. Also, we may need to serialize `SelectedRows` in the future, so putting serialize functions in util is file. Would you mind please update PR https://github.com/PaddlePaddle/Paddle/pull/5455 to add serialization functions?

Reply:
@helinwang 
We need to support multiple types, `LoDTensor`, `SelectedRows`, `LoDTensorArray`, we can put it as a template  free-function(global function). 
And I think:
1. `polymorphism` is not a real demand in application. When you do a serialization, you will have an object with a proper type. You can use a reference instead of a pointer.  The class always provide sufficient information to serialize itself.
http://www.boost.org/doc/libs/1_45_0/libs/serialization/doc/serialization.html

2. Other popular libraries always serialized object through free function. 
Such as protobuf, msgpack. 

```c++
MsgPack::Serializer serializer(socket);  
std::vector<std::unique_ptr<MsgPack::Element>> arrayWithoutElements, arrayWith3Elements;
arrayWith3Elements.push_back(MsgPack::Factory(true));
arrayWith3Elements.push_back(MsgPack__Factory(Array(std::move(arrayWithoutElements))));
arrayWith3Elements.push_back(MsgPack::Factory("Hello World!"));  
serializer << MsgPack__Factory(Array(std::move(arrayWith3Elements)));

MsgPack::Deserializer deserializer(socket);  
deserializer.deserialize([](std::unique_ptr<MsgPack::Element> parsed) {
    std::cout << "Parsed: " << *parsed << "\n";
    return false;
}, true);
```

Reply:
If we treat Tensor and the operators as a computing library, then we should use the free function. --- Just treat Tensor as a third-party library.
@typhoonzero @helinwang

Reply:
1. treat Tensor and the operators as a computing library, then we'll have to describe the data inside each type to be serialized. We are using `xxxDesc` for now, eg.
    ```
    message TensorDesc {
      required DataType data_type = 1;
      repeated int64 dims = 2; // [UNK, 640, 480] is saved as [-1, 640, 480]
    }
    ```
1. use the desc and the data pointers is enouph for a "Free Function" to serialize it.